### PR TITLE
Add a line breaker prototype by importing uax14_rs

### LINKED ISSUE #707: Add uax14_rs as an experiment crate
The issue is to discuss integrating https://github.com/makotokato/uax14_rs into icu4x as the first step toward #109. We may import it under `experiment/`, and improve it later.

Summary on uax14_rs status
---------------------
This may answer some of the questions in [this comment](https://github.com/unicode-org/icu4x/issues/109#issuecomment-832235347).

1. uax14_rs has implemented [spec_test.rs](https://github.com/makotokato/uax14_rs/blob/master/tests/spec_test.rs) that passes https://www.unicode.org/Public/UCD/latest/ucd/auxiliary/LineBreakTest.txt.
2. uax14_rs use [`generate_properties.py`](https://github.com/makotokato/uax14_rs/blob/master/tools/generate_properties.py) to generate line break properties and UAX14 rules tables, and use them to drive the finite state machine. See next section for its performance.

Performance comparison of line breaking implementations 
---------------
The line break benchmark is in https://github.com/zbraniecki/intl-measurements

The test is simple: read the data into a UTF-16 string, and pass to the line breaker. Measure the total time to iterate all the line break opportunities. Numbers are from [here](https://github.com/zbraniecki/intl-measurements/pull/7#issuecomment-831430959) and [here](https://github.com/zbraniecki/intl-measurements/pull/8#issue-630922163), running on Intel Core i5-9600K 6-Core 3.7 GHz, 32 GB RAM.

Lang / Impl | [uax14_rs](https://github.com/makotokato/uax14_rs) | icu4c (66) | [rust_icu](https://github.com/google/rust_icu)
--- | --- | --- | ---
English | 844ms | 1246ms | 1195ms
Chinese | 151ms | 389ms | 353ms
Thai [1] | 1687ms | 712ms |  716ms

[1] For Thai, uax14_rs uses  `experimental/segmenter_lstm/`; rust_icu and icu4c use dictionary.

Possible future work after the integration
-------------------
1. Use CodePointTrie to map codepoint to line break value. 
2. Use Rust instead of python to generate UAX14 line break rules into tables. 
3. Use `DataProvider` to plugin the rule table.
4. Improve documentation and tests.

Other thought? 

@makotokato Feel free to comment if I misinterpret anything, or add any potential future work you envision.

cc @jfkthame @zbraniecki

### PR DESCRIPTION
**Pull request by aethanyc**
Fixed #707.

This goal of this PR is to integrate uax14_rs with all the CI check pass. Any change to the code should be done in a follow up PR, e.g. fixing all the `cargo clippy` warnings.

### CODE REVIEW

**Comment by aethanyc** [comment]
@makotokato Could you also take a look at this PR?

File: experimental/segmenter/src/lstm.rs
Comment (aethanyc) [inline experimental/segmenter/src/lstm.rs:12]: Note: it probably makes more sense to put the data elsewhere so that it can be used by `segmenter` and `segmenter_lstm`. In the future, we might also want to add Burmese training data from https://github.com/unicode-org/lstm_word_segmentation.

The purpose of the change is not to add the same Thai data again into icu4x because it is large.

File: experimental/segmenter/src/lib.rs
Comment (sffc) [inline experimental/segmenter/src/lib.rs:13]: Suggestion: The Intl.Segmenter proposal has two steps here: first creating the Segmenter object, which takes the locale and performs data loading, and then having a `.segment()` method on it that returns the iterator for the string.  I think a similar API would be good to have here as well, because when we implement dynamic data loading, we'll want to have a Segmenter object with that data loaded into it.

Comment (aethanyc) [reply experimental/segmenter/src/lib.rs:13]: Agree that we should have a top level API similar to the Intl.Segmenter proposal once we implement other types of breaker.
It's a bit strange to add one for `LineBreakIterator` because current Intl.Segmenter doesn't have `granularity: "line"`.

Comment (sffc) [reply experimental/segmenter/src/lib.rs:13]: I'm thinking more about the data loading. In your API, where does the segmentation data reside?

Comment (aethanyc) [reply experimental/segmenter/src/lib.rs:13]: In @Makotokato's current design, there are two sets of data: the rule table from UAX 14 (`UAX14_RULE_TABLE`) and the table converting codepoint to line break property (`UAX14_PROPERTY_TABLE`). They reside statically as arrarys in [rule_table.rs](https://github.com/unicode-org/icu4x/blob/4b8b5bd5d8137293f6647ffb0e2b4d2529d55d62/experimental/segmenter/src/rule_table.rs#L9) and [property_table.rs](https://github.com/unicode-org/icu4x/blob/4b8b5bd5d8137293f6647ffb0e2b4d2529d55d62/experimental/segmenter/src/property_table.rs#L10), respectively.

A use case of `UAX14_PROPERTY_TABLE` when converting a codepoint to line break property.
https://github.com/unicode-org/icu4x/blob/4b8b5bd5d8137293f6647ffb0e2b4d2529d55d62/experimental/segmenter/src/line_breaker.rs#L48

Comment (sffc) [reply experimental/segmenter/src/lib.rs:13]: Do you plan on making those arrays loadable through the data provider, in order to ensure we always have fresh segmentation data?

Comment (aethanyc) [reply experimental/segmenter/src/lib.rs:13]: Yes, loading the rule tables through the DataProvider is on my todo list. I'll investigate it after this PR is merged. There are more todos in "Possible future work after the integration" https://github.com/unicode-org/icu4x/issues/707#issue-878294852

**Comment by aethanyc** [comment]
As of https://github.com/unicode-org/icu4x/commit/48c47db44cace4a608b627f6065de1dd39aac166, the CI requires a PR contains no clippy warnings, so I add commits to fix them.

File: experimental/segmenter/src/line_breaker.rs
Comment (Manishearth) [inline experimental/segmenter/src/line_breaker.rs:44]: suggestion: All of the conditions here seem to be running roughly the same code, seems like this can be simplified to just `let prop = get_linebreak_property_latin1(..);` followed by a single `if` condition that maps `CJ` to `ID`

File: experimental/segmenter/src/line_breaker.rs
Comment (Manishearth) [inline experimental/segmenter/src/line_breaker.rs:42]: question: what's this parameter for? can this be documented?

**Review by Manishearth — APPROVED** [review]
So overall I think this would be nicer if it were documented, but that might be a large ask of this PR which is just attempting to import this crate. At some point it's worth doing a second review once this is documented.

**Comment by aethanyc** [comment]
@Manishearth Thanks for the review. I've filed follow-up issues to address your comments.

@sffc Do you want to take another look at this PR? Is Manish's approval sufficient for this PR to be merged?

**Pull request merged**
